### DISCUSSION TITLE
Next new feature(s) ? Prioritise our - limited - resources toward your need

### DISCUSSION THREAD

**Discussion by gmaze** [euroargodev/argopy#214] [Ideas]
Since our dev team is pretty limited ([just check this out](https://github.com/euroargodev/argopy/graphs/contributors)), I'd like to prioritise resources toward the most urgent needs of users.
I tried to launched a 1 week poll in here on Twitter: https://twitter.com/mazeguillaume/status/1519931989297610754

If you have more suggestions, please list them below ...

**Comment by gmaze** [comment]
# Trajectory support
See current work about this subject: https://github.com/euroargodev/argopy/labels/argo-traj

Working on this would lead to:
- a new data access point for trajectory (path)
- interpolation for under-ice trajectory
- a new data set `traj` with the drifting data

**Reply by HCBScienceProducts** [reply]
To follow up from https://github.com/euroargodev/argopy/issues/361#issuecomment-2162592445, here three use cases with argo-traj files. 

* Case 1: Float behaviour diagnostics and analysis

Argo trajectory variables come in two groups: a N_MEASUREMENT dimension variable group and a N_CYCLE dimension variable group (-> see Argo User Manual). The N_MEASUREMENT group are always relative to one of the Primary Measurement Codes (any code evenly divisible by 100, e.g. 100, 200, 300, etc.) or one of the Secondary Measurement Codes (any code evenly divisible by 50 but not evenly divisible by 100, e.g. 150, 250, 450, etc.), where the Measurement Code gives information on the phase the float is currently in or transitioning to (see Reference Table 16 for details). N_MEASUREMENT and N_CYCLE groups are linked by the CYCLE_NUMBER(N_MEASUREMENT) variable.

To better understand the float behaviour, e.g., phases and durations of the park-and-profile cycles, the Measurement Code variables can be used. Attached an example for 6901929 of PRES vs. JULD colour-coded by Measurement Code (MC)  for March/April 2018 (cycles 5-8). On cycle 7, the float grounded on Deep Descent to the profile depth, which is traceable from the Argo trajectory.

![6901929_Rtraj_MC_series_illustration](https://github.com/user-attachments/assets/022b0894-245e-4a6b-a4b4-a59eb394bf3d)

Features this would require:
- [ ] access to N_MEASUREMENT variable group from trajectory
- [ ] Reference Table 16 labels and logic (relative codes) 
- [ ] Filter by Measurement Codes: (a) specific ones (e.g., MC=690, MC=703, MC=901), (b) group-wise (e.g., all relative to a given Primary/Secondary Measurement Code), or cross-group specific (e.g., all MC+11 where MC either 100, 600, 700, or 800)
- [ ] Time series plot of PRES vs. JULD with different color code options (e.g., Measurement Code like in example, Cycle Number, other N_MEASUREMENT variables like TEMP, ...)
- [ ] Add-on: Based on GPS-Positions and times (MC=703), extract depth from a bathymetry and add to time series plot

**Reply by HCBScienceProducts** [reply]
* Case 2: Extract observations during park drift, e.g., BBP

Some Argo floats perfom considerable data collections during park drift, e.g., useful for particle flux estimations from optical sensor data or for sensor QC/biofouling diagnostics (-> Poteau et al. 2017 work).

Features this would require
- [ ] Filter by Measurement Codes: Extract Park Drift data (everything towards MC 300: Park End Time, i.e., MC 275-324)
- [ ] Filter by PRES/depth horizon (e.g., if different park depths over float lifetime)

**Reply by HCBScienceProducts** [reply]
* Case 3: Oxygen optode in-air data and in-air calibration

To achieve highest DOXY accuracy, the SCOR WG142 recommended to systematically perform in-air observations by oxygen optodes, which has become best practice on all floats. These in-air and related measurements are stored in the Argo trajectory file.

A typical workflow would take the following steps:
1. Identify presence of oxygen in-air measurements, i.e., (a) PPOX_DOXY variable is present with (b) in-air related measurement codes and data (either MC+11 where MC 100, 600, 700, or 800 for dedicated in-air sequence, or MC-1 for opportunistic in-air measurements).
2. Find corresponding (nearest) profile cycle to extract GPS position and time (current cycle for MC 600, 700, 800 but _previous cycle_ for MC 100).
3. From an atmospheric reanalysis product, extract atmospheric pressure, air temperature and humidity for profile positions and times (e.g., via CDS API)
4. Identify presence of and extract corresponding oxygen/TEMP(/PSAL) in-water measurements either from the Argo trajectory from a dedicated sequence (MC+10 where MC 100, 600, 700, or 800), or from the corresponding profile files' DOXY profiles (e.g., shallowest observation >0.5 dbar)
5. Per profile/cycle, group oxygen partial pressure PPOX_DOXY in-air observations, calculate reanalysis-based oxygen partial pressure, and group/calculate in-water observations oxygen partial pressure.
6. From time series of 5., perform in-air analysis to determine oxygen gain slope factor (and time drift if present) (preferably using the common adjustment notation to present/output results)
7. Apply oxygen correction (preferably using the common adjustment notation) to data files (Argo b-prof, Argo traj).

Features this would require:
- [ ] Filter by Measurement Codes, e.g., all MC+11 where MC either 100, 600, 700, or 800
- [ ] Extract nearest sufacing (GPS position and time)
- [ ] Add-on: Interact with external API to get atmospheric reanalysis data based on GPS-Positions and times (MC=703)
- [ ] Group data by cycle number (e.g., 5 in-air measurements MC+11 per cycle, 5 in-water measurements MC+10 per cycle: group by median)
- [ ] Advanced: Match cycles with profile files, e.g., for near-surface TEMP, PSAL, DOXY extraction ("in-water" side if MC+10 is absent)
- [ ] Advanced: Add in-air analysis math and options and plots

**Reply by gmaze** [reply]
Thanks a lot @HCBScienceProducts for your input !

**Comment by gmaze** [comment]
# BGC data support
See current work about this subject: https://github.com/euroargodev/argopy/labels/argo-BGC

Working on this would lead to:
- full support for the data set `bgc`, i.e. access to BGC data in `expert` and `standard` user modes

**Comment by gmaze** [comment]
# Improved performances
See current issues about this subject: https://github.com/euroargodev/argopy/labels/performance
or discussions: https://github.com/euroargodev/argopy/discussions?discussions_q=label%3Aperformance

Working on this would lead to:
- support for the Euro-Argo dataselection webAPI
- caching at higher levels
- explore new data format, including native cloud storage (eg: zarr) or new [Akward arrays](https://github.com/euroargodev/argopy/discussions/210)

**Comment by gmaze** [comment]
# More visualisation tools
See current work about this subject: https://github.com/euroargodev/argopy/labels/plot

Working on this would lead to:
- [ ] Waterfall plots
- [ ] T/S diagram
- [ ] QC flags plots

**Reply by gmaze** [reply]
Some interesting new utilities for this in #245
